trainers/base_trainer.py

```diff
@@ -1,29 +1,155 @@
-from abc import abstractmethod
+import torch
+import numpy as np
+
+from torch import Tensor
+from torch.utils.data import DataLoader
+from torch.nn import Module, BCELoss
+from torch.optim import Optimizer, Adam, AdamW
+
+from tqdm import tqdm
+from loguru import logger
+from omegaconf.dictconfig import DictConfig
+
+from ..metric import (
+    precision_at_k, recall_at_k, map_at_k, ndcg_at_k
+)
 
 class BaseTrainer:
-    def __init__(self) -> None:
-        pass
+    def __init__(self, args: DictConfig) -> None:
+        self.args: DictConfig = args
+        self.device: torch.device = self._device(self.args.device)
+        self.model: Module = self._model(self.args.model).to(self.device)
+        self.optimizer: Optimizer = self._optimizer(self.args.optimizer, self.model, self.args.lr)
+        self.loss: BCELoss = self._loss(self.args.loss)
 
-    @abstractmethod
-    def run(self):
-        pass
+    def _device(self, device_name: str) -> torch.device:
+        if device_name.lower() in ('cpu', 'cuda',):
+            return torch.device(device_name.lower())
+        else:
+            logger.error(f"Not supported device: {device_name}")
+            return torch.device('cpu')
 
-    @abstractmethod
-    def train(self):
-        pass
+    def _model(self, model_name: str) -> Module:
+        if model_name.lower() in ('test',):
```

Comment:
After First implementation, we need to change this part to an implemented model.